python/ngraph/ops.py

```diff
@@ -633,6 +633,36 @@ def prod(node, reduction_axes=None, name=None):
     return Product(node, AxisSet(get_reduction_axes(node, reduction_axes)))
 
 
+@nameable_op
+def reduce(node,                 # type: Node
+           initial_value,        # type: ScalarData
+           reduction_node,       # type: Node
```

Comment:
This will need to be an nGraph `Function`. The idea is that you have some function on scalars `f(x,y) = ???`, and you're using that to reduce ("fold" in a functional programming sense) some of the axes. For example, if `f(x,y) = x+y`, and `n = {{1,2,3},{4,5,6}}` then (note: the following notation is pseudocode but it might be pretty close to the actual Python notation that falls out):

`reduce(n,0,f,reduction_axes=[1]) --> {0+1+2+3,0+4+5+6} --> {6,15}`

`reduce(n,0,f,reduction_axes=[0]) --> {0+1+4,0+2+5,0+3+6} --> {5,7,9}`

`reduce(n,0,f,reduction_axes=[0,1]) --> (0+1+2+3)+(0+4+5+6) --> 21`

_Except_ that I'm implying a higher degree of determinism here than actually exists: it's acceptable, and the backend may choose, to sum the columns first.

This can be used to implement more general reductions. For example, if `initial_value = 1` and `f(x,y) = x*y` then you can use it to take the product (of each row, or each column, or of an entire matrix, or of analogous things in higher dimensions). If `initial_value = -inf` and `f(x,y) = max(x,y)` you can take the maximum, and so on.

The reason it has to be a `Function` instead of a `Node` is that a `Node` itself can't be free-standing if it takes inputs---it has to be hooked up to some other node(|s) that generate(s|) its input(|s). So even if we are doing a simple `Add`, that `Add` still has to be wrapped up in a nested `Function`, whose inputs are represented as `Parameter` nodes. Think of it like the `lambda` that shows up in this example: http://book.pythontips.com/en/latest/map_filter.html#reduce .

Just as an aside, note that this op made its way into nGraph originally because we needed it in order to implement this op from XLA:

https://www.tensorflow.org/performance/xla/operation_semantics#reduce

In practice, we tend to avoid its use (favoring instead ops like `Sum` or `Max`) because it is hard to compile it efficiently in the fully general case. (There are ways it can be done, but for the moment implementing them is a low priority.)